### PR TITLE
Add support for Waveshare EPD 2.13" V3

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

Add support for the [V3 Waveshare 2.13" e-paper display](https://www.waveshare.com/wiki/Pico-ePaper-2.13).

While this is possibly the same display as the 2.13in-ttgo-dke already supported this PR provides improved performance - faster updates, less blocking of the loop thread and improved partial updates.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3171

## Test Environment

- [x] ESP32
- [x] ESP32 IDF

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
time:
  - platform: sntp
    id: ntp

spi:
  mosi_pin: GPIO4
  miso_pin: GPIO5
  clk_pin: GPIO6

display:
  - platform: waveshare_epaper
    rotation: 270
    cs_pin: GPIO7
    dc_pin: GPIO15
    busy_pin: GPIO16
    reset_pin: GPIO17
    model: 2.13inv3
    full_update_every: 30
    lambda: |-
      it.strftime(70, 100, id(roboto), TextAlign::BASELINE_LEFT, "%H:%M:%S", id(ntp).now());
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: tests/test4.yaml [hunk 1]
```diff
@@ -572,6 +572,15 @@ display:
     powerup_pin: GPIO1
     wakeup_pin: GPIO1
     vcom_pin: GPIO1
+  - platform: waveshare_epaper
+    cs_pin: GPIO23
```
Comment (clydebarrow) [inline tests/test4.yaml]:
```suggestion
    spi_id: spi_id_1
    cs_pin: GPIO23
```

File: esphome/components/waveshare_epaper/waveshare_epaper.h [hunk 2]
```diff
@@ -49,7 +50,7 @@ class WaveshareEPaper : public display::DisplayBuffer,
       this->reset_pin_->digital_write(false);
       delay(reset_duration_);  // NOLINT
       this->reset_pin_->digital_write(true);
-      delay(200);  // NOLINT
+      delay(20);  // NOLINT
```
Comment (jesserockz) [inline esphome/components/waveshare_epaper/waveshare_epaper.h]:
```suggestion
      delay(20);
```

Comment (clydebarrow) [reply esphome/components/waveshare_epaper/waveshare_epaper.h]: ~~I wrote this some time - I'm older and wiser now. Will rework to use `set_timeout()`.~~
Done.

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp [hunk 3]
```diff
@@ -2262,7 +2275,7 @@ void HOT WaveshareEPaper2P13InDKE::display() {
     this->start_data_();
     this->write_array(this->buffer_, this->get_buffer_length_());
     this->end_data_();
-    delay(300);  // NOLINT
+    // delay(300);  // NOLINT
```
Comment (jesserockz) [inline esphome/components/waveshare_epaper/waveshare_epaper.cpp]: Lets just remove it =)
```suggestion
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp (hunk 3 above)
Comment (clydebarrow) [reply esphome/components/waveshare_epaper/waveshare_epaper.cpp]: Well, I tested without it and it works, on my display at least. That code is used by other displays, but the `wait_until_idle()` should take care of things.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**